README
NAME
    Audio::Scan - Fast C metadata and tag reader for all common audio file
    formats

SYNOPSIS
        use Audio::Scan;

        my $data = Audio::Scan->scan('/path/to/file.mp3');

        # Just file info
        my $info = Audio::Scan->scan_info('/path/to/file.mp3');

        # Just tags
        my $tags = Audio::Scan->scan_tags('/path/to/file.mp3');
    
        # Scan without reading (possibly large) artwork into memory.
        # Instead of binary artwork data, the size of the artwork will be returned instead.
        {
            local $ENV{AUDIO_SCAN_NO_ARTWORK} = 1;
            my $data = Audio::Scan->scan('/path/to/file.mp3');
        }
    
        # Scan a filehandle
        open my $fh, '<', 'my.mp3';
        my $data = Audio::Scan->scan_fh( mp3 => $fh );
        close $fh;

DESCRIPTION
    Audio::Scan is a C-based scanner for audio file metadata and tag
    information. It currently supports MP3, MP4, Ogg Vorbis, FLAC, ASF, WAV,
    AIFF, Musepack, Monkey's Audio, and WavPack.

    See below for specific details about each file format.

METHODS
  scan( $path )
    Scans $path for both metadata and tag information. The type of scan
    performed is determined by the file's extension. Supported extensions
    are:

        MP3:  mp3, mp2
        MP4:  mp4, m4a, m4b, m4p, m4v, m4r, k3g, skm, 3gp, 3g2, mov
        AAC (ADTS): aac
        Ogg:  ogg, oga
        FLAC: flc, flac, fla
        ASF:  wma, wmv, asf
        Musepack:  mpc, mpp, mp+
        Monkey's Audio:  ape, apl
        WAV: wav
        AIFF: aiff, aif
        WavPack: wv

    This method returns a hashref containing two other hashrefs: info and
    tags. The contents of the info and tag hashes vary depending on file
    format, see below for details.

  scan_info( $path )
    If you only need file metadata and don't care about tags, you can use
    this method.

  scan_tags( $path )
    If you only need the tags and don't care about the metadata, use this
    method.

  scan_fh( $type => $fh )
    Scans a filehandle. $type is the type of file to scan as, i.e. "mp3" or
    "ogg". Note that FLAC does not support reading from a filehandle.

  find_frame( $path, $timestamp_in_ms )
    Returns the byte offset to the first audio frame starting from the given
    timestamp (in milliseconds).

    MP3, Ogg, FLAC, ASF, MP4
        The byte offset to the data packet containing this timestamp will be
        returned. For file formats that don't provide timestamp information
        such as MP3, the best estimate for the location of the timestamp
        will be returned. This will be more accurate if the file has a Xing
        header or is CBR for example.

    WAV, AIFF, Musepack, Monkey's Audio, WavPack
        Not yet supported by find_frame.

  find_frame_return_info( $mp4_path, $timestamp_in_ms )
    The header of an MP4 file contains various metadata that refers to the
    structure of the audio data, making seeking more difficult to perform.
    This method will return the usual $info hash with 2 additional keys:

        seek_offset - The seek offset in bytes
        seek_header - A rewritten MP4 header that can be prepended to the audio data
                      found at seek_offset to construct a valid bitstream. Specifically,
                      the following boxes are rewritten: stts, stsc, stsz, stco

    For example, to seek 30 seconds into a file and write out a new MP4 file
    seeked to this point:

        my $info = Audio::Scan->find_frame_return_info( $file, 30000 );
    
        open my $f, '<', $file;
        sysseek $f, $info->{seek_offset}, 1;

        open my $fh, '>', 'seeked.m4a';
        print $fh $info->{seek_header};

        while ( sysread( $f, my $buf, 65536 ) ) {
            print $fh $buf;
        }

        close $f;
        close $fh;

  find_frame_fh( $type => $fh, $offset )
    Same as "find_frame", but with a filehandle.

  has_flac()
    Deprecated. Always returns 1 now that FLAC is always enabled.

  is_supported( $path )
    Returns 1 if the given path can be scanned by Audio::Scan, or 0 if not.

  get_types()
    Returns an array of strings of the file types supported by Audio::Scan.

  extensions_for( $type )
    Returns an array of strings of the file extensions that are considered
    to be the file type *$type*.

  type_for( $extension )
    Returns file type for a given extension. Returns *undef* for unsupported
    extensions.

MP3
  INFO
    The following metadata about a file may be returned:

        id3_version (i.e. "ID3v2.4.0")
        song_length_ms (duration in milliseconds)
        layer (i.e. 3)
        stereo
        samples_per_frame
        padding
        audio_size (size of all audio frames)
        audio_offset (byte offset to first audio frame)
        bitrate (in bps, determined using Xing/LAME/VBRI if possible, or average in the worst case)
        samplerate (in kHz)
        vbr (1 if file is VBR)

        If a Xing header is found:
        xing_frames
        xing_bytes
        xing_quality

        If a VBRI header is found:
        vbri_delay
        vbri_frames
        vbri_bytes
        vbri_quality

        If a LAME header is found:
        lame_encoder_version
        lame_tag_revision
        lame_vbr_method
        lame_lowpass
        lame_replay_gain_radio
        lame_replay_gain_audiophile
        lame_encoder_delay
        lame_encoder_padding
        lame_noise_shaping
        lame_stereo_mode
        lame_unwise_settings
        lame_source_freq
        lame_surround
        lame_preset

  TAGS
    Raw tags are returned as found. This means older tags such as ID3v1 and
    ID3v2.2/v2.3 are converted to ID3v2.4 tag names. Multiple instances of a
    tag in a file will be returned as arrays. Complex tags such as APIC and
    COMM are returned as arrays. All tag fields are converted to upper-case.
    All text is converted to UTF-8.

    Sample tag data:

        tags => {
              ALBUMARTISTSORT => "Solar Fields",
              APIC => [ "image/jpeg", 3, "", <binary data snipped> ],
              CATALOGNUMBER => "INRE 017",
              COMM => ["eng", "", "Amazon.com Song ID: 202981429"],
              "MUSICBRAINZ ALBUM ARTIST ID" => "a2af1f31-c9eb-4fff-990c-c4f547a11b75",
              "MUSICBRAINZ ALBUM ID" => "282143c9-6191-474d-a31a-1117b8c88cc0",
              "MUSICBRAINZ ALBUM RELEASE COUNTRY" => "FR",
              "MUSICBRAINZ ALBUM STATUS" => "official",
              "MUSICBRAINZ ALBUM TYPE" => "album",
              "MUSICBRAINZ ARTIST ID" => "a2af1f31-c9eb-4fff-990c-c4f547a11b75",
              "REPLAYGAIN_ALBUM_GAIN" => "-2.96 dB",
              "REPLAYGAIN_ALBUM_PEAK" => "1.045736",
              "REPLAYGAIN_TRACK_GAIN" => "+3.60 dB",
              "REPLAYGAIN_TRACK_PEAK" => "0.892606",
              TALB => "Leaving Home",
              TCOM => "Magnus Birgersson",
              TCON => "Ambient",
              TCOP => "2005 ULTIMAE RECORDS",
              TDRC => "2004-10",
              TIT2 => "Home",
              TPE1 => "Solar Fields",
              TPE2 => "Solar Fields",
              TPOS => "1/1",
              TPUB => "Ultimae Records",
              TRCK => "1/11",
              TSOP => "Solar Fields",
              UFID => [
                    "http://musicbrainz.org",
                    "1084278a-2254-4613-a03c-9fed7a8937ca",
              ],
        },

MP4
  INFO
    The following metadata about a file may be returned:

        audio_offset (byte offset to start of mdat)
        compatible_brands
        file_size
        leading_mdat (if file has mdat before moov)
        major_brand
        minor_version
        song_length_ms
        timescale
        tracks (array of tracks in the file)
            Each track may contain:
        
            audio_type
            avg_bitrate
            bits_per_sample
            channels
            duration
            encoding
            handler_name
            handler_type
            id
            max_bitrate
            samplerate

  TAGS
    Tags are returned in a hash with all keys converted to upper-case. Keys
    starting with 0xA9 (copyright symbol) will have this character stripped
    out. Sample tag data:

        tags => {
           AART              => "Album Artist",
           ALB               => "Album",
           ART               => "Artist",
           CMT               => "Comments",
           COVR              => <binary data snipped>,
           CPIL              => 1,
           DAY               => 2009,
           DESC              => "Video Description",
           DISK              => "1/2",
           "ENCODING PARAMS" => "vers\0\0\0\1acbf\0\0\0\2brat\0\1w\0cdcv\0\1\6\5",
           GNRE              => "Jazz",
           GRP               => "Grouping",
           ITUNNORM          => " 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000",
           ITUNSMPB          => " 00000000 00000840 000001E4 00000000000001DC 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000",
           LYR               => "Lyrics",

Changes
Revision history for Audio::Scan

Note: Bug numbers refer to bugs at http://bugs.slimdevices.com

1.08	2023-11-29
	- mpeg-4: Smoothly handle small and ignore empty atoms.
	- Added Windows build support for Strawberry Perl 32/64-bit.

1.07	Unreleased build
	- ogg/opus: Improve/fix seek.

1.06	2022-11-10
	- opus: Fix parsing large comment headers (such as large embedded images)

1.05    2021-09-10
	- ID3: Fix v2.4 extended header handling.
	- WavPack DSD: fix song_length_ms calculation (Kimmo Taskinen).

1.04    2021-08-09
        - MP4: Allow seek with sample rates > 65535 (32 bits)
        - Fixed missing zlib reference in generated Makefile.
	- Link zlib based on OS.

1.02    2018-10-16
        - Opus: song_length_ms and bitrate_average were not always scanned properly

1.01    2018-07-09
        - Added Opus codec support. (Jeff Muizelaar)
        - ADTS: RT #124525, fixed incorrect duration calcuation due to off-by-one error
          in frame counter.
        - Added missing license meta name. (manwar)

1.00    2018-04-21
        - Fixed parsing of DSF files with more than 2 channels.

0.99    2017-11-24
        - Fixed a bug where an incorrect audio offset was returned when using
          NO_ARTWORK and reading a file with an ID3v2.2 or v2.3 unsynchronization bit
          set. The artwork in this type of file is only available when scanning in
          normal mode (without NO_ARTWORK).
        - Added id3_was_unsynced => 1 to the info section for v2.2/v2.3 files with the
          whole-tag unsync bit set. This type of tag is generally harmful to parser performance,
          so this flag might be useful if you want to find and upgrade this type of tag to v2.4.
        - DSF: WAV_BLOCK_SIZE was being used instead of DSF_BLOCK_SIZE. (Kimmo Taskinen)
        - ID3: Support for reading ID3 tags located >4GB into a file, such as in very large
          DSF files. (Kimmo Taskinen)
        - WavPack: Read correct samplerate and bits_per_sample for DSD files. (Kimmo Taskinen)
        - DSF/DFF: added bitrate to info hash.

0.98    2017-04-28
        - RT #119101, stop including MYMETA files in the tarball.

0.97    2017-04-27
        - WAV: GH #2, fixed an integer overflow error that could result in an incorrect song_length_ms
          for >16-bit files.

0.96    2016-11-21
        - Ogg: RT #118888, fixed crash when trying to read Vorbis comments in an
          incorrectly framed file with a bad terminal header page.
        - Removed hints/darwin.pl, don't try to be smarter than Perl about how to build things.
        - Fixed various minor compiler warnings output by the latest clang on macOS.

0.95    Unreleased community build
        - Improved DFF parsing for files which end with corrupted chunk. (Adrian Smith)
        - Improved ID3 support for DFF files. (Adrian Smith)

0.94    Unreleased community build
        - DSF/DFF: Support dsd file scanning for dsd and dsdiff files. (Kimmo Taskinen)
        - AIFF: Fixed scanning of 24 bit files. (Adrian Smith)

0.93    2011-09-14
        - ASF: Support seeking in a file without an ASF_Index object.
        - Don't use MAX_PATH because this limits path names to only 260 chars on Win32.

0.92    2011-09-09
        - Added $info->{jenkins_hash} which is a 32-bit hash value of the
          filename + mtime + file size.
        - MP4: Get correct HE-AAC samplerate values from esds.
        - ADTS AAC: Add support for .adts file extension.

0.91    2011-09-06
        - MP3, MP4, WMA, WAV, AIFF, ADTS AAC: Added DLNA audio profile detection as
          $info->{dlna_profile}.

0.90    2011-08-03
        - ID3: Bug 17392, fixed APIC offset calculation for non-MP3 files such as AIFF
          and WAV where the tag does not live at the front of the file.

0.89    2011-08-03
        - APE: Bug 15895, better handling if we have a broken APE tag and run out of
          tag data. (Adrian Smith)
        - WMA: Bug 17355, fixed WM/Picture offset calculation when artwork is stored in the
          Header Extension/Metadata Library block.
        - WMA: Added output warning when trying to seek in a file without ASF_Index.
        - Switched aac.t to Test::Warn instead of manual STDERR redirection.

0.88    2011-07-20
        - ADTS AAC: Check first 2-3 frames to avoid false syncs.
        - OSX: Xcode 4 doesn't include the 10.5 SDK, so check for this before using it.
        - Win32: Display better error messages when unable to read from files.
        - MP3, AAC, MPC: Bug 17173: Cleaned up several places where invalid memory reads could occur.


src/ogg.c
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "ogg.h"

int
get_ogg_metadata(PerlIO *infile, char *file, HV *info, HV *tags)
{
  return _ogg_parse(infile, file, info, tags, 0);
}

int
_ogg_parse(PerlIO *infile, char *file, HV *info, HV *tags, uint8_t seeking)
{
  Buffer ogg_buf, vorbis_buf;
  unsigned char *bptr;
  unsigned int buf_size;

  unsigned int id3_size = 0; // size of leading ID3 data

  off_t file_size;           // total file size
  off_t audio_size;          // total size of audio without tags
  off_t audio_offset = 0;    // offset to audio

  unsigned char ogghdr[OGG_HEADER_SIZE];
  char header_type;
  int serialno;
  int final_serialno;
  int pagenum;
  uint8_t num_segments;
  int pagelen;
  int page = 0;
  int packets = 0;
  int streams = 0;

  unsigned char vorbishdr[23];
  unsigned char channels;
  unsigned int blocksize_0 = 0;
  unsigned int avg_buf_size;
  unsigned int samplerate = 0;
  unsigned int bitrate_nominal = 0;
  uint64_t granule_pos = 0;

  unsigned char vorbis_type = 0;

  int i;
  int err = 0;

  buffer_init(&ogg_buf, OGG_BLOCK_SIZE);
  buffer_init(&vorbis_buf, 0);

  file_size = _file_size(infile);
  my_hv_store( info, "file_size", newSVuv(file_size) );

  if ( !_check_buf(infile, &ogg_buf, 10, OGG_BLOCK_SIZE) ) {
    err = -1;
    goto out;
  }

  // Skip ID3 tags if any
  bptr = (unsigned char *)buffer_ptr(&ogg_buf);
  if (
    (bptr[0] == 'I' && bptr[1] == 'D' && bptr[2] == '3') &&
    bptr[3] < 0xff && bptr[4] < 0xff &&
    bptr[6] < 0x80 && bptr[7] < 0x80 && bptr[8] < 0x80 && bptr[9] < 0x80
  ) {
    /* found an ID3 header... */
    id3_size = 10 + (bptr[6]<<21) + (bptr[7]<<14) + (bptr[8]<<7) + bptr[9];

    if (bptr[5] & 0x10) {
      // footer present
      id3_size += 10;
    }

    buffer_clear(&ogg_buf);

    audio_offset += id3_size;

    DEBUG_TRACE("Skipping ID3v2 tag of size %d\n", id3_size);

    PerlIO_seek(infile, id3_size, SEEK_SET);
  }

  while (1) {
    // Grab 28-byte Ogg header
    if ( !_check_buf(infile, &ogg_buf, OGG_HEADER_SIZE, OGG_BLOCK_SIZE) ) {
      err = -1;
      goto out;
    }

    buffer_get(&ogg_buf, ogghdr, OGG_HEADER_SIZE);

    audio_offset += OGG_HEADER_SIZE;

    // check that the first four bytes are 'OggS'
    if ( ogghdr[0] != 'O' || ogghdr[1] != 'g' || ogghdr[2] != 'g' || ogghdr[3] != 'S' ) {
      PerlIO_printf(PerlIO_stderr(), "Not an Ogg file (bad OggS header): %s\n", file);
      goto out;
    }

    // Header type flag
    header_type = ogghdr[5];

    // Absolute granule position, used to find the first audio page
    bptr = ogghdr + 6;
    granule_pos = (uint64_t)CONVERT_INT32LE(bptr);
    bptr += 4;
    granule_pos |= (uint64_t)CONVERT_INT32LE(bptr) << 32;

    // Stream serial number
    serialno = CONVERT_INT32LE((ogghdr+14));

    // Count start-of-stream pages
    if ( header_type & 0x02 ) {
      streams++;
    }

    // Keep track of packet count
    if ( !(header_type & 0x01) ) {
      packets++;
    }

    // stop processing if we reach the 3rd packet and have no data
    if (packets > 2 * streams && !buffer_len(&vorbis_buf) ) {
      break;
    }

    // Page seq number
    pagenum = CONVERT_INT32LE((ogghdr+18));

    if (page >= 0 && page == pagenum) {
      page++;
    }
    else {
      page = -1;
      DEBUG_TRACE("Missing page(s) in Ogg file: %s\n", file);
    }

    DEBUG_TRACE("OggS page %d / packet %d at %d\n", pagenum, packets, (int)(audio_offset - OGG_HEADER_SIZE));
    DEBUG_TRACE("  granule_pos: %llu\n", granule_pos);

    // If the granule_pos > 0, we have reached the end of headers and
    // this is the first audio page
    if (granule_pos > 0 && granule_pos != -1) {
      // If seeking, don't waste time on comments
      if (seeking) {
        break;
      }

      // Parse comments, but only if we have any extra data in the buffer
      if ( buffer_len(&vorbis_buf) > 0 ) {
        _parse_vorbis_comments(infile, &vorbis_buf, tags, 1);
        DEBUG_TRACE("  parsed vorbis comments\n");
      }

      buffer_clear(&vorbis_buf);

      break;
    }

    // Number of page segments
    num_segments = ogghdr[26];

    // Calculate total page size
    pagelen = ogghdr[27];
    if (num_segments > 1) {
      int i;

      if ( !_check_buf(infile, &ogg_buf, num_segments, OGG_BLOCK_SIZE) ) {
        err = -1;
        goto out;
      }

      for( i = 0; i < num_segments - 1; i++ ) {
        u_char x;
        x = buffer_get_char(&ogg_buf);
        pagelen += x;
      }

      audio_offset += num_segments - 1;
    }

    if ( !_check_buf(infile, &ogg_buf, pagelen, OGG_BLOCK_SIZE) ) {
      err = -1;
      goto out;
    }

    // Still don't have enough data, must have reached the end of the file
    if ( buffer_len(&ogg_buf) < pagelen ) {
      PerlIO_printf(PerlIO_stderr(), "Premature end of file: %s\n", file);

      err = -1;
      goto out;
    }

    audio_offset += pagelen;

    // Copy page into vorbis buffer
    buffer_append( &vorbis_buf, buffer_ptr(&ogg_buf), pagelen );
    DEBUG_TRACE("  Read %d into vorbis buffer\n", pagelen);

    // Process vorbis packet
    if ( !vorbis_type ) {
      vorbis_type = buffer_get_char(&vorbis_buf);
      // Verify 'vorbis' string
      if ( strncmp( buffer_ptr(&vorbis_buf), "vorbis", 6 ) ) {
        PerlIO_printf(PerlIO_stderr(), "Not a Vorbis file (bad vorbis header): %s\n", file);
        goto out;
      }
      buffer_consume( &vorbis_buf, 6 );

      DEBUG_TRACE("  Found vorbis packet type %d\n", vorbis_type);
    }

    if (vorbis_type == 1) {
      // Parse info
      // Grab 23-byte Vorbis header
      if ( buffer_len(&vorbis_buf) < 23 ) {
        PerlIO_printf(PerlIO_stderr(), "Not a Vorbis file (bad vorbis header): %s\n", file);
        goto out;
      }

      buffer_get(&vorbis_buf, vorbishdr, 23);

      my_hv_store( info, "version", newSViv( CONVERT_INT32LE(vorbishdr) ) );

      channels = vorbishdr[4];
      my_hv_store( info, "channels", newSViv(channels) );
      my_hv_store( info, "stereo", newSViv( channels == 2 ? 1 : 0 ) );

      samplerate = CONVERT_INT32LE((vorbishdr+5));
      my_hv_store( info, "samplerate", newSViv(samplerate) );
      my_hv_store( info, "bitrate_upper", newSViv( CONVERT_INT32LE((vorbishdr+9)) ) );

      bitrate_nominal = CONVERT_INT32LE((vorbishdr+13));
      my_hv_store( info, "bitrate_nominal", newSViv(bitrate_nominal) );
      my_hv_store( info, "bitrate_lower", newSViv( CONVERT_INT32LE((vorbishdr+17)) ) );

      blocksize_0 = 2 << ((vorbishdr[21] & 0xF0) >> 4);
      my_hv_store( info, "blocksize_0", newSViv( blocksize_0 ) );
      my_hv_store( info, "blocksize_1", newSViv( 2 << (vorbishdr[21] & 0x0F) ) );

      DEBUG_TRACE("  parsed vorbis info header\n");

      buffer_clear(&vorbis_buf);
      vorbis_type = 0;
    }

    // Skip rest of this page
    buffer_consume( &ogg_buf, pagelen );
  }

  buffer_clear(&ogg_buf);

  // audio_offset is 28 less because we read the Ogg header
  audio_offset -= OGG_HEADER_SIZE;

  // from the first packet past the comments
  my_hv_store( info, "audio_offset", newSViv(audio_offset) );

  audio_size = file_size - audio_offset;
  my_hv_store( info, "audio_size", newSVuv(audio_size) );

  my_hv_store( info, "serial_number", newSVuv(serialno) );

  // calculate average bitrate and duration
  avg_buf_size = blocksize_0 * 2;
  if ( file_size > avg_buf_size ) {
    DEBUG_TRACE("Seeking to %d to calculate bitrate/duration\n", (int)(file_size - avg_buf_size));
    PerlIO_seek(infile, file_size - avg_buf_size, SEEK_SET);
  }
  else {
    DEBUG_TRACE("Seeking to %d to calculate bitrate/duration\n", (int)audio_offset);
    PerlIO_seek(infile, audio_offset, SEEK_SET);
  }

  if ( PerlIO_read(infile, buffer_append_space(&ogg_buf, avg_buf_size), avg_buf_size) == 0 ) {
    if ( PerlIO_error(infile) ) {
      PerlIO_printf(PerlIO_stderr(), "Error reading: %s\n", strerror(errno));
    }
    else {
      PerlIO_printf(PerlIO_stderr(), "File too small. Probably corrupted.\n");
    }

    err = -1;
    goto out;
  }

  // Find sync
  bptr = (unsigned char *)buffer_ptr(&ogg_buf);
  buf_size = buffer_len(&ogg_buf);
  while (
    buf_size >= 14
    && (bptr[0] != 'O' || bptr[1] != 'g' || bptr[2] != 'g' || bptr[3] != 'S')
  ) {
    bptr++;
    buf_size--;

    if ( buf_size < 14 ) {
      // Give up, use less accurate bitrate for length
      DEBUG_TRACE("buf_size %d, using less accurate bitrate for length\n", buf_size);

      my_hv_store( info, "song_length_ms", newSVpvf( "%d", (int)((audio_size * 8) / bitrate_nominal) * 1000) );
      my_hv_store( info, "bitrate_average", newSViv(bitrate_nominal) );

      goto out;
    }
  }
  bptr += 6;

  // Get absolute granule value
  granule_pos = (uint64_t)CONVERT_INT32LE(bptr);
  bptr += 4;
  granule_pos |= (uint64_t)CONVERT_INT32LE(bptr) << 32;
  bptr += 4;

  // Get serial number of this page, if the serial doesn't match the beginning of the file
  // we have changed logical bitstreams and can't use the granule_pos for bitrate
  final_serialno = CONVERT_INT32LE((bptr));

  if ( granule_pos && samplerate && serialno == final_serialno ) {
    // XXX: needs to adjust for initial granule value if file does not start at 0 samples
    int length = (int)((granule_pos * 1.0 / samplerate) * 1000);
    my_hv_store( info, "song_length_ms", newSVuv(length) );
    my_hv_store( info, "bitrate_average", newSVuv( _bitrate(audio_size, length) ) );

    DEBUG_TRACE("Using granule_pos %llu / samplerate %d to calculate bitrate/duration\n", granule_pos, samplerate);
  }
  else {
    // Use nominal bitrate
    my_hv_store( info, "song_length_ms", newSVpvf( "%d", (int)((audio_size * 8) / bitrate_nominal) * 1000) );
    my_hv_store( info, "bitrate_average", newSVuv(bitrate_nominal) );

    DEBUG_TRACE("Using nominal bitrate for average\n");
  }

out:
  buffer_free(&ogg_buf);
  buffer_free(&vorbis_buf);

  if (err) return err;

  return 0;
}

void
_parse_vorbis_comments(PerlIO *infile, Buffer *vorbis_buf, HV *tags, int has_framing)
{
  unsigned int len;
  unsigned int num_comments;
  char *tmp;
  char *bptr;
  SV *vendor;

  // Vendor string
  len = buffer_get_int_le(vorbis_buf);
  vendor = newSVpvn( buffer_ptr(vorbis_buf), len );
  sv_utf8_decode(vendor);
  my_hv_store( tags, "VENDOR", vendor );
  buffer_consume(vorbis_buf, len);

  // Number of comments
  num_comments = buffer_get_int_le(vorbis_buf);

  while (num_comments--) {
    len = buffer_get_int_le(vorbis_buf);

    // Sanity check length
    if ( len > buffer_len(vorbis_buf) ) {
      DEBUG_TRACE("invalid Vorbis comment length: %u\n", len);
      return;
    }

    bptr = buffer_ptr(vorbis_buf);

    if (
#ifdef _MSC_VER
      !strnicmp(bptr, "METADATA_BLOCK_PICTURE=", 23)
#else
      !strncasecmp(bptr, "METADATA_BLOCK_PICTURE=", 23)
#endif
    ) {
      // parse METADATA_BLOCK_PICTURE according to http://wiki.xiph.org/VorbisComment#METADATA_BLOCK_PICTURE
      AV *pictures;
      HV *picture;
      Buffer pic_buf;
      uint32_t pic_length;

      buffer_consume(vorbis_buf, 23);

      // Copy picture into new buffer and base64 decode it
      buffer_init(&pic_buf, len - 23);
      buffer_append( &pic_buf, buffer_ptr(vorbis_buf), len - 23 );
      buffer_consume(vorbis_buf, len - 23);

      _decode_base64( buffer_ptr(&pic_buf) );

      picture = _decode_flac_picture(infile, &pic_buf, &pic_length);
      if ( !picture ) {
        PerlIO_printf(PerlIO_stderr(), "Invalid Vorbis METADATA_BLOCK_PICTURE comment\n");
      }
      else {
        DEBUG_TRACE("  found picture of length %d\n", pic_length);

        if ( my_hv_exists(tags, "ALLPICTURES") ) {
          SV **entry = my_hv_fetch(tags, "ALLPICTURES");
          if (entry != NULL) {
            pictures = (AV *)SvRV(*entry);
            av_push( pictures, newRV_noinc( (SV *)picture ) );
          }
        }
        else {
          pictures = newAV();

          av_push( pictures, newRV_noinc( (SV *)picture ) );

          my_hv_store( tags, "ALLPICTURES", newRV_noinc( (SV *)pictures ) );
        }
      }

      buffer_free(&pic_buf);
    }
    else if (
#ifdef _MSC_VER
      !strnicmp(bptr, "COVERART=", 9)
#else
      !strncasecmp(bptr, "COVERART=", 9)
#endif
    ) {
      // decode COVERART into ALLPICTURES
      AV *pictures;
      HV *picture = newHV();

      // Fill in recommended default values for most of the picture hash
      my_hv_store( picture, "color_index", newSVuv(0) );
      my_hv_store( picture, "depth", newSVuv(0) );
      my_hv_store( picture, "description", newSVpvn("", 0) );
      my_hv_store( picture, "height", newSVuv(0) );
      my_hv_store( picture, "width", newSVuv(0) );
      my_hv_store( picture, "mime_type", newSVpvn("image/", 6) ); // As recommended, real mime should be in COVERARTMIME
      my_hv_store( picture, "picture_type", newSVuv(0) ); // Other

      if ( _env_true("AUDIO_SCAN_NO_ARTWORK") ) {
        my_hv_store( picture, "image_data", newSVuv(len - 9) );
        buffer_consume(vorbis_buf, len);
      }
      else {
        int pic_length;

        buffer_consume(vorbis_buf, 9);
        pic_length = _decode_base64( buffer_ptr(vorbis_buf) );
        DEBUG_TRACE("  found picture of length %d\n", pic_length);

        my_hv_store( picture, "image_data", newSVpvn( buffer_ptr(vorbis_buf), pic_length ) );
        buffer_consume(vorbis_buf, len - 9);
      }

      if ( my_hv_exists(tags, "ALLPICTURES") ) {
        SV **entry = my_hv_fetch(tags, "ALLPICTURES");
        if (entry != NULL) {
          pictures = (AV *)SvRV(*entry);
          av_push( pictures, newRV_noinc( (SV *)picture ) );
        }
      }
      else {
        pictures = newAV();

        av_push( pictures, newRV_noinc( (SV *)picture ) );

        my_hv_store( tags, "ALLPICTURES", newRV_noinc( (SV *)pictures ) );
      }
    }
    else {
      New(0, tmp, (int)len + 1, char);
      buffer_get(vorbis_buf, tmp, len);
      tmp[len] = '\0';

      _split_vorbis_comment( tmp, tags );

      Safefree(tmp);
    }
  }

  if (has_framing) {
    // Skip framing byte (Ogg only)
    buffer_consume(vorbis_buf, 1);
  }
}

static int
ogg_find_frame(PerlIO *infile, char *file, int offset)
{
  int frame_offset = -1;
  uint32_t samplerate;
  uint32_t song_length_ms;
  uint64_t target_sample;

  // We need to read all metadata first to get some data we need to calculate
  HV *info = newHV();
  HV *tags = newHV();
  if ( _ogg_parse(infile, file, info, tags, 1) != 0 ) {
    goto out;
  }

  song_length_ms = SvIV( *(my_hv_fetch( info, "song_length_ms" )) );
  if (offset >= song_length_ms) {
    goto out;
  }

  samplerate = SvIV( *(my_hv_fetch( info, "samplerate" )) );

  // Determine target sample we're looking for
  target_sample = (uint64_t)offset * samplerate / 1000;
  DEBUG_TRACE("Looking for target sample %llu\n", target_sample);

  frame_offset = _ogg_binary_search_sample(infile, file, info, target_sample);

out:
  // Don't leak
  SvREFCNT_dec(info);
  SvREFCNT_dec(tags);

  return frame_offset;
}

int
_ogg_binary_search_sample(PerlIO *infile, char *file, HV *info, uint64_t target_sample)
{
  Buffer buf;
  unsigned char *bptr;
  unsigned int buf_size;
  int frame_offset = -1;
  int best_frame_offset = -1;
  uint64_t granule_pos = 0;
  off_t low;
  off_t high;

  off_t audio_offset = SvIV( *(my_hv_fetch( info, "audio_offset" )) );
  off_t file_size    = SvIV( *(my_hv_fetch( info, "file_size" )) );
  uint32_t serialno  = SvIV( *(my_hv_fetch( info, "serial_number" )) );

  // Define the binary search range
  low  = audio_offset;
  high = file_size - OGG_HEADER_SIZE;

  buffer_init(&buf, OGG_MAX_PAGE_SIZE + OGG_HEADER_SIZE);

  while (high > low) {
    off_t mid;
    off_t page_start_offset = -1;
    uint32_t cur_serialno;
    int i; // Used by macro CONVERT_INT32LE

    if (high - low > OGG_HEADER_SIZE) {
      mid = low + ((high - low) / 2);
    } else {
      mid = low; // Fast-forward search
    }

    DEBUG_TRACE("  Searching for sample %llu between %d and %d (mid %d)\n", target_sample, low, high, mid);

    if (mid > file_size - OGG_HEADER_SIZE) {
      DEBUG_TRACE("  Reached end of file, aborting\n");
      frame_offset = -1;
      goto out;
    }

    if (PerlIO_seek(infile, mid, SEEK_SET) == -1) {
      frame_offset = -1;
      goto out;
    }

    buffer_clear(&buf);

    // Worst case is:
    // ....OggS...<OGG_MAX_PAGE_SIZE>...OggS
    //      ^-mid                        ^-high
    //
    // To handle this, read OGG_HEADER_SIZE bytes extra after 'high'
    // so that we find the header that starts just before 'high'.
    if (!_check_buf(infile, &buf, OGG_HEADER_SIZE,
		    MIN(OGG_MAX_PAGE_SIZE, high - mid) + OGG_HEADER_SIZE)) {
      frame_offset = -1;
      goto out;
    }

    for (bptr = buffer_ptr(&buf), buf_size = buffer_len(&buf); ; ++bptr, --buf_size) {
      if (buf_size < 4) {
	// no page start found!?
	break;
      }

      if (bptr[0] != 'O' || bptr[1] != 'g' || bptr[2] != 'g' || bptr[3] != 'S') {
	continue;
      }

      page_start_offset = buffer_len(&buf) - buf_size;
      frame_offset = mid + page_start_offset;
      break;
    }

    if (page_start_offset < 0) {
      DEBUG_TRACE("  Nothing found in upper half, searching lower\n");
      high = mid;
      continue;
    }

    DEBUG_TRACE("  checking frame at %d\n", frame_offset);

    // Read granule_pos for this packet
    bptr = buffer_ptr(&buf);
    bptr += page_start_offset + 6;
    granule_pos = (uint64_t)CONVERT_INT32LE(bptr);
    bptr += 4;
    granule_pos |= (uint64_t)CONVERT_INT32LE(bptr) << 32;
    bptr += 4;

    // Also read serial number, if this ever changes within a file it is a chained
    // file and we can't seek
    cur_serialno = CONVERT_INT32LE(bptr);
    if (serialno != cur_serialno) {
      DEBUG_TRACE("  serial number changed to %x, aborting seek\n", cur_serialno);
      frame_offset = -1;
      goto out;
    }

    if (granule_pos > target_sample) {
      best_frame_offset = frame_offset;
      DEBUG_TRACE("  searching lower\n");
      high = mid;
    }
    else if (granule_pos < target_sample) {
      DEBUG_TRACE("  searching higher\n");
      low = frame_offset + OGG_HEADER_SIZE;
    }
    else {
      DEBUG_TRACE("  found frame at %d\n", frame_offset);
      best_frame_offset = frame_offset;
      break;
    }
  }

  frame_offset = best_frame_offset;

out:
  buffer_free(&buf);
  return frame_offset;
}
